go/database/mpt/io/parallel_visit.go

```diff
@@ -0,0 +1,429 @@
+
+	// open all sources for the number of workers and track possible errors
+	sources := make([]NodeSource, 0, NumWorker)
+	openingErrors := make([]error, 0, NumWorker)
+	for i := 0; i < NumWorker; i++ {
+		source, err := sourceFactory.Open()
+		sources = append(sources, source)
+		openingErrors = append(openingErrors, err)
+	}
+
+	defer func() {
+		// only log errors when closing the sources
+		// as data are opened parallel to the main program
+		// and thus the error should not invalidate the result
+		for _, source := range sources {
+			if source != nil {
+				if err := source.Close(); err != nil {
+					log.Printf("failed to close source: %v", err)
+				}
+			}
+		}
+	}()
+
+	// error in one of the sources interrupts processing
+	if err := errors.Join(openingErrors...); err != nil {
+		return err
+	}
+
+	for i := 0; i < NumWorker; i++ {
+		go func(source NodeSource) {
+			for !done.Load() {
+				// TODO: implement throttling
+				// get the next job
+				requestsMutex.Lock()
+				req, present := requests.Pop()
+				requestsMutex.Unlock()
+
+				// process the request
+				if !present {
+					continue
+				}
+
+				// fetch the node and put it into the responses
+				node, err := source.Get(req.id)
+
+				responsesMutex.Lock()
+				responses[req.id] = response{node, err}
+				responsesCond.Signal()
+				responsesMutex.Unlock()
+
+				// if there was a fetch error, stop the workers
+				if err != nil {
+					done.Store(true)
+					return
+				}
+
+				// derive child nodes to be fetched
+				switch node := node.(type) {
+				case *mpt.BranchNode:
+					children := node.GetChildren()
+					requestsMutex.Lock()
+					for i, child := range children {
+						id := child.Id()
+						if id.IsEmpty() {
+							continue
+						}
+						pos := req.position.Child(byte(i))
+						requests.Add(request{pos, child.Id()})
+					}
+					requestsMutex.Unlock()
+				case *mpt.ExtensionNode:
+					next := node.GetNext()
+					requestsMutex.Lock()
+					pos := req.position.Child(0)
+					requests.Add(request{pos, next.Id()})
+					requestsMutex.Unlock()
+				case *mpt.AccountNode:
+					if !cutAtAccounts {
+						storage := node.GetStorage()
+						id := storage.Id()
+						if !id.IsEmpty() {
+							requestsMutex.Lock()
+							pos := req.position.Child(0)
+							requests.Add(request{pos, id})
+							requestsMutex.Unlock()
+						}
+					}
+				}
+			}
+		}(sources[i])
+	}
+
+	// Perform depth-first iteration through the trie.
+	stack := []mpt.NodeId{root}
+	for len(stack) > 0 {
+		cur := stack[len(stack)-1]
+		stack = stack[:len(stack)-1]
+
+		var res response
+		responsesMutex.Lock()
+		for {
+			found := false
+			res, found = responses[cur]
+			if found {
+				delete(responses, cur)
+				break
+			}
+			responsesCond.Wait()
+		}
+		responsesMutex.Unlock()
+
+		if res.err != nil {
+			return res.err
+		}
+
+		switch visitor.Visit(res.node, mpt.NodeInfo{Id: cur}) {
+		case mpt.VisitResponseAbort:
+			return nil
+		case mpt.VisitResponsePrune:
```

Comment:
I do not think that this algorithm is supporting visitors deciding to prune the visiting process. This should be removed and documented -- or even the type of visitor accepted by this function should be changed.

The reason is: if a visitor decides to "Prune" a part of the tree, the workers do not know about this. They will still fetch all the nodes of that part of the tree, and those nodes are not consumed. This can easily end up in a OOM crash.

Reply:
ok, I created a new type of visitor. Also I needed to create a visitor to wrap the `cutAtAccounts` (`pruneStorage`) flag, i.e. it is a bit complex at the moment. Let me know what do you think.